osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/GetKeysService.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2021 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.application.services;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.SecurityKeyType;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.GetKeysRequestDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.GetKeysResponseDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.KeyDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.SecretTypeDto;
+import org.opensmartgridplatform.shared.security.RsaEncrypter;
+import org.opensmartgridplatform.ws.schema.core.secret.management.SecretType;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GetKeysService {
+
+  private final SecretManagementService secretManagementService;
+  private final RsaEncrypter rsaEncrypter;
+
+  @Autowired
+  public GetKeysService(
+      final SecretManagementService secretManagementService, final RsaEncrypter rsaEncrypter) {
+    this.secretManagementService = secretManagementService;
+    this.rsaEncrypter = rsaEncrypter;
+  }
+
+  public GetKeysResponseDto getKeys(
+      final DlmsDevice device, final GetKeysRequestDto getKeysRequestDto) {
+
+    final List<KeyDto> keys =
+        this.getEncryptedKeys(device.getDeviceIdentification(), getKeysRequestDto.getsecretTypes());
```

Comment:
This solution (looking ahead in the current implementation behind this call) leads to multiple SOAP calls to the secret management service (if multiple secret types are requested).
It looks like this overhead could be avoided by first converting the secret types from the `getKeysRequestDto` to a list of `SecurityKeyType` and then calling `secretManagementService.getKeys(deviceIdentification, securityKeyTypes)` which would retrieve the keys in a single call to `secretManagementClient.getSecretsRequest`.

Reply:
Good catch! I've updated the code.